pep-0682.rst

```diff
@@ -0,0 +1,183 @@
+of mathematics, negative zero is surprising or unwanted -- especially
+in the context of displaying an (often rounded) numerical result.  This PEP
+proposes an extension to the `string format specification`_, allowing negative
+zero to be normalized to positive zero.
+
+.. _`signed zero`: https://en.wikipedia.org/wiki/Signed_zero
+.. _`string format specification`: https://docs.python.org/3/library/string.html#formatstrings
+
+
+Motivation
+==========
+
+Here is negative zero:
+
+.. code-block:: pycon
+
+    >>> x = -0.
+    >>> x
+    -0.0
+
+When formatting a number, negative zero can result from rounding.  Assuming
+the user's intention is truly to discard precision, the distinction between
+negative and positive zero of the rounded result might be considered an
+unwanted artifact:
+
+.. code-block:: pycon
+
+    >>> for x in (.002, -.001, .060):
+    ...     print(f'{x: .1f}')
+     0.0
+    -0.0
+     0.1
+
+There are various approaches to clearing the sign of a negative zero.  It
+can be achieved without a conditional by adding positive zero:
+
+.. code-block:: pycon
+
+    >>> x = -0.
+    >>> x + 0.
+    0.0
+
+To normalize negative zero when formatting, it is necessary to perform
+a redundant (and error-prone) pre-rounding of the input:
+
+.. code-block:: pycon
+
+    >>> for x in (.002, -.001, .060):
+    ...     print(f'{round(x, 1) + 0.: .1f}')
+     0.0
+     0.0
+     0.1
+
+What we would like instead is a first-class option to normalize negative
+zero, on top of everything else that numerical string formatting already
+offers.
+
+
+Rationale
+=========
+
+There are use cases where negative zero is unwanted in formatted number
+output -- arguably it's the more common case.  Expanding the format
+specification is the best way to support this because number formatting
+already incorporates rounding, and the normalization of negative zero must
+happen after rounding.
+
+While it is possible to pre-round and normalize a number before formatting,
+it's tedious and prone to error if the rounding doesn't precisely match
+that of the format spec.  Furthermore, functions that wrap formatting would
+find themselves having to parse format specs to extract the precision
+information.  For example, consider how this utility for formatting
+one-dimensional numerical arrays would be complicated by such pre-rounding:
+
+.. code-block:: python
+
+    def format_vector(v, format_spec='8.2f'):
+        """Format a vector (any iterable) using given per-term format string."""
+        return f"[{','.join(f'{term:{format_spec}}' for term in v)}]"
+
+The solution must be opt-in, because we can't change the behavior of
+programs that may be expecting or relying on negative zero when formatting
+numbers.
+
+The proposed extension is intentionally ``[sign][z]`` rather than
+``[sign[z]]``.  The default for ``sign`` (``-``) is not widely known or
+explicitly written, so this avoids everyone having to learn it just to use
+the ``z`` option.
```

Comment:
This is a matter of draughting, so you're well within your rights as author to ignore. It does feel a slight jump in topic though -- the previous paragraphs discuss the format specification in the round, and why the solution must be invisible to those not using it, etc. This paragraph though skips detail of the current format specification or grammar of the mini-langauge itself, going straight into the nested vs sibiling approach. 

A possible solution would be to combine this paragraph with the preceeding, and say something along the lines of the solution must be opt in, and given that the negative option is the default we can't link the two, moving the detail of the grammar to the specification.

Another solution is to move this entire paragraph to the specification -- this could be argued to be explanation of the choices in the specification rather than rationale for the chage itself.

Reply:
Thank you.  This is related to the discussion with CAM on section ordering: https://github.com/python/peps/pull/2295#discussion_r798041177.  I originally wrote the rationale section intending that it support the decisions of the specification, and it was placed after the specification.

> this could be argued to be explanation of the choices in the specification rather than rationale for the change itself.

As CAM cited from PEP 1:

> "The rationale fleshes out the specification by describing why particular design decisions were made."

This suggests that the rational supports the specification, and is not merely a rationale for the change itself.  If it does support the specification, logically it should follow that section.  I wish that the guidelines on section ordering gave me such leeway.

Possible resolutions, with my preference first:
   1. move Rationale after Specification (perhaps naming it "Specification Rationale" for clarity)
   2. move some rationale text into the specification, as you suggest.  (My view is that it makes the specification less concise, and I was happy to detail these things in another section.)

Reply:
Thanks for the extracts, I missed that discussion.

My own opinion is that clarity of expression is more important than adherence to the "rules". (That said, the "rules" exist for a reason!). I've also never been entirely sure of the difference between Motivation and Rationale.

I'd say flip the order if it makes the most sense. 

@CAM-Gerlach -- would you have a strong objection to this?

A

Reply:
My actual suggestion was to not include this paragraph, the one following, and optionally the one preceding it in the Rationale at all:

> Finally, you can make your last couple paragraphs separate subsections in the Rejected Ideas section (and maybe the opt-out too), since they each discuss why a particular alternative was not adopted, along with any you might add based on my suggestions above, and any others suggest.

They are rather detailed and implementation-level (as @AA-Turner noticed), and refer to why specific potential alternatives were not adopted, which is exactly what belongs in the Rejected Ideas section (with subheadings for each separate idea)—which, indeed, goes _after_ the Specification, where you want them.

This just leaves the current mention of prior art (and any you add), the brief high-level rationale of the core design decision, and the counterpoint to why pre-rounding instead wouldn't work (the latter of which, if you really wanted, you _could_ consider moving to Motivation or as a separate rejected idea), which makes much more sense as an introduction to the specification. If you prefer, you could also consider re-organizing those sub-parts a bit to put the pre-rounding approach first (why the status quo doesn't work), prior art section second (how did other languages tackle the problem) and the high-level description of your design last, which then smoothy outlines the rationale that led to the overall idea of the proposal, which then leads into the specification itself.

Reply:
> (2) the reasons for choosing the particular solution that's being proposed from amongst other possible solutions.

As I've been mentioning, the high level reasoning that led to the solution goes in the Rationale section, but the specific reasons why other possible solutions were not selected, and those dealing with the details of the specification, belong in the Rejected Ideas section, which goes _after_ the Specification section. Is there a reason you'd rather not use that section for the more specific alternatives and details of the specification, while still keeping a short high-level rationale and comparison of prior art (which, certainly, should go before proposing what we're doing, its "prior" art after all) before? That seems like a good compromise that moves the stuff that you and Adam identified that doesn't really belong before the specification, after it, while still keeping the stuff that leads into it before and following PEP 1 and PEP 12 on the Rationale section itself?

Reply:
> Is there a reason you'd rather not use that section for the more specific alternatives and details of the specification

No - I'll step out of this and leave the wrangling to you, Adam and John. I don't really care much about the exact order of the sections or whether there's strict adherence to PEP 12 or not - I care much more that the resulting text is coherent and is meaningful when read from top to bottom.

Reply:
I also don't have a strong opinion, having said my piece above!

> I don't really care much about the exact order of the sections or whether there's strict adherence to PEP 12 or not - I care much more that the resulting text is coherent and is meaningful when read from top to bottom.

Seconded, although perhaps verboten for a pep editor to say :P

A

Reply:
>  the high level reasoning that led to the solution goes in the Rationale section, but the specific reasons why other possible solutions were not selected, and those dealing with the details of the specification, belong in the Rejected Ideas section

I would use the Rejected Ideas section if there was an idea that had legs and was actually rejected.  To apply an opposite polarity to every design decision and render it as a rejected idea seems a little contrived.  The items currently documented in Rationale, for the most part, had no reasonable alternatives-- and I touch on why that is for each.  A (fictitious) example of something I'd put into Rejected Ideas would be "We explored automatically inferring suppression of negative zero based on whether the input was rounded, but it was found to not be reliable and still may break existing programs."

I appreciate that the PEP guidelines are an established thing, and I'm merely a PEP author.  On the other hand, my eyes are fresh to this, and there may be ways to refine the guidelines.  At the least, I've come upon some contradiction in the documentation as to whether Rational is that of the entire PEP or of the Specification.

What I'd like to do from here is keep the current text, but as a "Specification Rationale" subsection within "Specification", so as to make the aim clear.  I hope it's an acceptable compromise, given that the sanctioned "Rationale" section is optional.

The attention from everyone towards making a good document has been impressive and welcome-- even if we have different views on it-- thank you.

Reply:
I think I've addressed what Adam raised originally by moving relevant text into Specification.